heads.go

```diff
@@ -4,36 +4,51 @@ import (
 	"bytes"
 	"encoding/binary"
 	"errors"
-	"sort"
-	"strings"
-
 	cid "github.com/ipfs/go-cid"
 	ds "github.com/ipfs/go-datastore"
 	"github.com/ipfs/go-datastore/query"
 	dshelp "github.com/ipfs/go-ipfs-ds-help"
 	logging "github.com/ipfs/go-log/v2"
+	"sort"
+	"strings"
+	"sync"
 )
 
 // heads manages the current Merkle-CRDT heads.
 type heads struct {
-	store     ds.Datastore
+	store ds.Datastore
+	// cache contains the current contents of the store
+	cache     map[cid.Cid]uint64
+	cacheMux  sync.RWMutex
 	namespace ds.Key
 	logger    logging.StandardLogger
 }
 
-func newHeads(store ds.Datastore, namespace ds.Key, logger logging.StandardLogger) *heads {
-	return &heads{
+func newHeads(store ds.Datastore, namespace ds.Key, logger logging.StandardLogger) (*heads, error) {
+	hh := &heads{
 		store:     store,
 		namespace: namespace,
 		logger:    logger,
+		cache:     make(map[cid.Cid]uint64),
+	}
+	if err := hh.primeCache(); err != nil {
+		return nil, err
 	}
+	return hh, nil
 }
 
 func (hh *heads) key(c cid.Cid) ds.Key {
 	// /<namespace>/<cid>
 	return hh.namespace.Child(dshelp.MultihashToDsKey(c.Hash()))
 }
 
+func canonicalizeCid(c cid.Cid) (cid.Cid, error) {
+	// Can we no-op this if c is already a V1 PB CID?
+	key := dshelp.MultihashToDsKey(c.Hash())
+	return dshelp.DsKeyToCidV1(key, cid.DagProtobuf)
+}
```

Comment:
I think this is not needed. Heads are always going to be in the same CID format and we don't want the conversion perf impact.